src/lib/formats/rpk.h

```diff
@@ -0,0 +1,145 @@
+// license:LGPL-2.1+
+// copyright-holders:Michael Zapf
+/***************************************************************************
+
+	rpk.h
+
+	RPK format support
+
+***************************************************************************/
+
+#ifndef MAME_FORMATS_RPK_H
+#define MAME_FORMATS_RPK_H
+
+#pragma once
+
+#include "hash.h"
+#include "unzip.h"
+
+#include <list>
+#include <optional>
+
+namespace util {
+/***************************************************************************
+	TYPE DEFINITIONS
+***************************************************************************/
+
+class rpk_reader;
+class rpk_file;
+
+// ======================> rpk_socket
+
+class rpk_socket
+{
+	friend class rpk_reader;
+public:
+	enum class socket_type
+	{
+		ROM,
+		RAM,
+		PERSISTENT_RAM
+	};
+
+	// ctor/dtor
+	rpk_socket(rpk_file &rpk, std::string &&id, socket_type type, std::string &&filename, std::optional<hash_collection> &&hashes, std::optional<std::uint32_t> length);
+	rpk_socket(const rpk_socket &) = delete;
+	rpk_socket(rpk_socket &&) = delete;
+	~rpk_socket();
+
+	// accessors
+	const std::string &id() const noexcept { return m_id; }
+	socket_type type() const noexcept { return m_type; }
+	const std::string &filename() const noexcept { return m_filename; }
+	std::uint32_t length() const noexcept { return m_length.value(); }
```

Comment:
"length" is required for sockets of type RAM, but optional for ROM (the size of the file drives the length for ROMs).  So it would be illegal to invoke `length()` is `type()` is not `socket_type::RAM`

Should the code be changed to be an assert and the std::optional<> usage removed to make this more obvious?

https://www.ninerpedia.org/wiki/MESS_cartridge_handling

Reply:
Oh, I see – it’s actually enforced that it’s set for RAM type on loading.  Well, it’s still better to assert and `return *m_length;` to avoid the overhead of the extra stuff that calling `.value()` does in that case.